### DISCUSSION TITLE
Env Variables and Modes

### DISCUSSION THREAD

**Discussion by damvik** [quasarframework/quasar#15303] [General - Components / Directives / etc] [answered]
Hello,
I am trying to use env variables based on [vite documentation](https://vitejs.dev/guide/env-and-mode.html). I need multiple envs and there is example how to use that:

> By default, the dev server (dev command) runs in development mode and the build command runs in production mode.

```bash
vite build --mode staging
```

But when I use --mode=staging flag in `quasar build` command i have an error  `App • ⚠️  Unknown mode "staging"`. I know quasar has same option so there is a conflict. How to use different mode only for Vite? Or should I handle process envs differently?

_EDIT:_
I found [this solution](https://github.com/quasarframework/quasar/discussions/10234) with cross-env, but this one with Vite is working good, and I think this should work in Quasar too.

**Comment by MilosPaunovic** [comment]
I've recently written this, seems like it is exactly what you need, so give it a look:
https://gist.github.com/MilosPaunovic/1da784b83466197196b4b0fd6448c0b1

**Comment by damvik** [comment]
@MilosPaunovic This is great solution, but I'm curious if that vite mode can work properly in Quasar. Vite env mode works out of the box and we don't need to install additional packages to handle it. So if there is any chance we can change mode in vite it should work without any problem.

**Reply by MilosPaunovic** [reply]
I'm not sure I understand the issue you're having. Could you try to explain it a bit more? Also, take a look at this https://quasar.dev/quasar-cli-vite/handling-process-env

**Reply by damvik** [reply]
I'm trying to use env variables without dotenv and cross-env. As you can see [here](https://quasar.dev/quasar-cli-vite/handling-process-env) in docs is provied link to Vite documentation. There is [explanation](https://vitejs.dev/guide/env-and-mode.html#modes) how to use multiple envs with mode option. 

![image](https://user-images.githubusercontent.com/58105519/214540335-6d1b4034-afbd-47c3-911a-8d499f986568.png)
![image](https://user-images.githubusercontent.com/58105519/214540294-f1533de3-bf3b-4ec8-bf12-2a55458d5e84.png)

If you read Vite docs you will understand what I mean.

**Reply by MilosPaunovic** [reply]
Yes, you can use those variables out-of-the box.

**Reply by damvik** [reply]
Yes but I can't change mode (To be precise I don't know how to do it). So I can use only .env.development and .env.production when Vite handle processing envs. Are you understand now what is a problem?

**Reply by MilosPaunovic** [reply]
That `mode` you're referring to is the mode of the application itself, and it could be one of the values form the circled list. Development and production are values of the `NODE_ENV` variable. And, if you want your own environments, you're gonna have to create them yourself.

**Reply by damvik** [reply]
@MilosPaunovic Did you read `vite` docs? I provided link [here](https://github.com/quasarframework/quasar/discussions/15303#discussioncomment-4775833). As I said previously there is `mode` option in `vite` bundler. So if we use this bundler in Quasar we should have access to define `mode` option for `vite`. I don't know how to say it more clearly

**Reply by rotu** [reply]
This is confusing because Vite defines e.g. `process.env.MODE: "electron"` whereas vite defines `import.meta.MODE: "development"` and these are both different from `process.env["MODE"]`, which is the runtime variable of the environment variable `$MODE`

**Comment by BenJackGill** [comment]
Did you ever get this working out of the box with Vite?

The custom solution from @MilosPaunovic looks cool but I'd rather stick with a native Vite solution if possible.

**Reply by damvik** [reply]
> Did you ever get this working out of the box with Vite?

I think it's impossible in Quasar right now

**Reply by rstoenescu** [reply]
Yarn/Npm install dotenv and in your quasar.config.js just import it then supply process.env to build > env.

**Reply by damvik** [reply]
@rstoenescu thank you for your response but this is not a solution to this problem. As you can see in previous comments we are trying to achieve something else. We want to use env variables based on Vite bulit-in .env. 

When we want use different .env file we need to call this command:

```bash
vite build --mode staging
```

Problem is with `--mode` flag, Quasar is using it in different way so there some conflict.

**Comment by rstoenescu** [comment, marked as answer]
Implemented out of the box support for .env files, as follows:

```
Order matters!

# quasarMode can be: spa | pwa | electron | capacitor | cordova | bex | ssr

.env                                                       # loaded in all cases
.env.local                                              # loaded in all cases, ignored by git
.env.[dev|prod]                                    # loaded for dev or prod only
.env.local.[dev|prod]                           # loaded for dev or prod only, ignored by git
.env.[quasarMode]                              # loaded for specific Quasar CLI mode only
.env.local.[quasarMode]                     # loaded for specific Quasar CLI mode only, ignored by git
.env.[dev|prod].[quasarMode]           # loaded for specific Quasar CLI mode and dev|prod only
.env.local.[dev|prod].[quasarMode]  # loaded for specific Quasar CLI mode and dev|prod only, ignored by git
```

Also, new quasar.config file props:

```
  /**
   * Folder where Quasar CLI should look for .env* files.
   * Can be an absolute path or a relative path to project root directory.
   *
   * @default project root directory
   */
  build > envFolder?: string;
  /**
   * Additional .env* files to be loaded.
   * Each entry can be an absolute path or a relative path to quasar.config > build > envFolder.
   *
   * @example ['.env.somefile', '../.env.someotherfile']
   */
  build > envFiles?: string[];
```

Will be available in next q/app-vite and q/app-webpack (version to be decided).

There is no "staging" support (at least not yet) in Quasar CLI.

**Reply by rstoenescu** [reply]
It's now available. Released q/app-vite v2.0.0-beta.1 and q/app-webpack v4.0.0-beta.1

**Reply by ricar2ce** [reply]
don't work for me 🤷‍♂️🤷‍♂️

**Reply by nicobao** [reply]
@rstoenescu Do we have to setup `build > envFolder?: string;` for it to work, and download dotenv (`yarn add dotenv`)?
I was not able to get it to work as well, with or without the above.
Is it possible to add the .env setup in the `yarn create quasar` template?
Also, I noticed that this page does not mention this change at all: https://quasar.dev/quasar-cli-vite/handling-process-env/ so it's a little confusing

**Reply by rstoenescu** [reply]
Just add one (or multiple) of those dotfiles in your root project. You will get this banner when files are detected:

<img width="813" alt="image" src="https://github.com/user-attachments/assets/3b8adf52-7196-45bd-ae1d-9760d46b7d09">

Should you want to place the dotfiles in some other place, only then use the quasar.config > build > envFolder setting.

Also, please note that there is no watcher for these files. If you add/remove/change these files, you will need to kill restart your devserver (kill then call "quasar dev" again).

**Reply by rstoenescu** [reply]
Also, you don't need to install anything else (like dotenv yourself).

**Reply by jclaveau** [reply]
I just opened a PR to be able to customize modes in the `envFiles` entry of the config file https://github.com/quasarframework/quasar/pull/17960
